dask/array/core.py

```diff
@@ -3873,13 +3872,93 @@ def tile(A, reps):
     return concatenate(reps * [A], axis=-1)
 
 
+def elemwise_array_mask(op, *args, **kwargs):
+    """
+    Modified version of :func:`elemwise` for getitem where the mask is
+    a boolean array, which requires slightly different reduction and
+    broadcasting rules from other operations.
+
+    See also
+    --------
+    atop
+    elemwise
+    slice_with_dask_array
+    """
+    if not set(['name', 'dtype']).issuperset(kwargs):
+        msg = "%s does not take the following keyword arguments %s"
+        raise TypeError(msg % (op.__name__, str(sorted(set(kwargs) - set(['name', 'dtype'])))))
+    args = [np.asarray(a) if isinstance(a, (list, tuple)) else a for a in args]
+    arrays = [a for a in args if not is_scalar_for_elemwise(a)]
+
+    # these are the two differences from elemwise
+    out_ndim = arrays[0].ndim
+    atop_args = concat((a, tuple(range(out_ndim)[::-1])[:a.ndim]) for a in arrays)
+
+    expr_inds = tuple(range(out_ndim))[::-1]
+    dt = arrays[0].dtype
+
+    name = kwargs.get('name', None) or '%s-%s' % (funcname(op),
+                                                  tokenize(op, dt, *args))
+    return atop(op, expr_inds, *atop_args, dtype=dt, name=name)
```

Comment:
I suspect it would probably fail spectacularly if used with anything other than `slice_with_dask_arrray`. That function is already a bit long, so I'll make it `elemwise_array_mask` private.